<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">


# Introduction to Pandas



Pandas is the most popular Python package for managing data sets. It's used extensively by data scientists.

### Learning Objectives

- Define the anatomy of DataFrames.
- Explore data with DataFrames.

### Lesson Guide

- [Introduction to `pandas`](#introduction)
- [Loading CSV Files](#loading_csvs)
- [Exploring Your Data](#exploring_data)
- [Data Dimensions](#data_dimensions)
- [DataFrames vs. Series](#dataframe_series)
- [Using the `.info()` Function](#info)
- [Using the `.describe()` Function](#describe)
- [Independent Practice](#independent_practice)
- [Pandas Indexing](#indexing)
- [Creating DataFrames](#creating_dataframes)
- [Checking Data Types](#dtypes)
- [Renaming and Assignment](#renaming_assignment)
- [Logical Filtering](#filtering)
- [Sorting](#sorting)
- [Review](#review)

<a id='introduction'></a>

### What is a dataframe?

---
The concept of a "dataframe" comes from the world of statistical software used in empirical research; 
- Generally refers to "tabular" data: a data structure representing cases (rows), each of which consists of a number of observations or measurements (columns)
- **Each row** is treated as a **single observation** of **multiple "variables"** 
- The row ("record") datatype can be **heterogenous** (a tuple of different types) 
- The column datatype must be **homogenous**. 
- Data frames usually contain some **metadata** in addition to **data**; for example, column and row names (unlike Numpy by default)

<a id='introduction'></a>

### What is `pandas`?

---

- A data analysis library — **P**anel **D**ata **S**ystem.
- It was created by Wes McKinney and open sourced by AQR Capital Management, LLC in 2009.
- It's implemented in highly optimized Python/Cython.
- It's the **most ubiquitous tool** used to start data analysis projects within the Python scientific ecosystem.


### Pandas Use Cases

---

- Cleaning data/munging.
- Exploratory analysis.
- Structuring data for plots or tabular display.
- Joining disparate sources.
- Modeling.
- Filtering, extracting, or transforming. 


![](https://snag.gy/tpiLCH.jpg)

![](https://snag.gy/1V0Ol4.jpg)

### Common Outputs

---

With `pandas` you can:

- Export to databases
- Integrate with `matplotlib`
- Collaborate in common formats (plus a variety of others)
- Integrate with Python built-ins (**and `numpy`!**)


### Importing `pandas`

---

Import `pandas` at the top of your notebook like so:

In [1]:
import pandas as pd
import numpy as np

Recall that the **`import pandas as pd`** syntax nicknames the `pandas` module as **`pd`** for convenience.

<a id='loading_csvs'></a>

### Loading a CSV into a DataFrame

---

`pandas` can load many types of files, but one of the most commonly used for storing data is a ```.csv```. As an example, let's load a data set on diamond characteristics from the ```./datasets``` directory:

In [2]:
diamonds = pd.read_csv('datasets/diamonds.csv')

This creates a `pandas` object called a **DataFrame**. DataFrames are powerful containers, featuring many built-in functions for exploring and manipulating data.

We will barely scratch the surface of DataFrame functionality in this lesson, but, throughout this course, you will become an expert at using them.

In short, a dataframe is a supercharged 2D array:
    - it has the data
    - it has information about it (meta-data - like columns names, etc...)

<a id='exploring_data'></a>

### Exploring Data using DataFrames

---

DataFrames come with built-in functionality that makes data exploration easy. 

To start, let's look at the **"header"** of your data using the ```.head()``` function. If run alone in a notebook cell, it will show you the first handful of columns in the data set, along with the first five rows.

In [3]:
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

If we want to see the last part of our data, we can use the ```.tail()``` function equivalently.

In [4]:
diamonds.tail()

carat        cut color clarity  depth  table  price     x     y     z
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64

In [7]:
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

<a id='data_dimensions'></a>

### Data Dimensions

---

It's always good to look at the dimensions of your data. The ```.shape``` property will tell you how many rows and columns are contained within your DataFrame.

In [8]:
diamonds.shape

(53940, 10)

As you can see, we have 53940 rows and 10 columns.

You'll also notice that this function operates the same as `.shape` for `numpy` arrays/matricies. **`pandas` makes use of **numpy** under its hood** for optimization and speed.

You can look up the names of your columns using the ```.columns``` property.


In [9]:
diamonds.columns

Index(['carat', 'cut', 'color', 'clarity', 'depth', 'table', 'price', 'x', 'y',
       'z'],
      dtype='object')

Accessing a specific column is easy. You can use **bracket** syntax just like you would with **Python dictionaries**, using the column's string name to extract it.

In [10]:
diamonds['carat'].head()

0    0.23
1    0.21
2    0.23
3    0.29
4    0.31
Name: carat, dtype: float64

As you can see, we can also use the ```.head()``` function on a single column, which is represented as a `pandas` Series object.

With a **list of strings**, you can also access a column (as a DataFrame instead of a Series).

In [11]:
diamonds[['carat']].head()

carat
0   0.23
1   0.21
2   0.23
3   0.29
4   0.31

In [17]:
diamonds[['color','carat','price',]].head(2)

color  carat  price
0     E   0.23    326
1     E   0.21    326

<a id='dataframe_series'></a>

### DataFrame vs. Series

---

There is an important difference between using a list of strings versus only using a string with a column's name: When you use a list containing the string, it returns another **DataFrame**. But, when you only use the string, it returns a `pandas` **Series** object.

In [19]:
print(type(diamonds['cut']))

<class 'pandas.core.series.Series'>


In [20]:
print(type(diamonds[['cut']]))

<class 'pandas.core.frame.DataFrame'>


**Breakout (2min):** What's the difference between `pandas` Series and DataFrame objects?

As long as your column names **don't contain any spaces** or other specialized characters (underscores are OK), you can access a column as a property of a DataFrame.  

**Get in the habit of referencing your Series columns using `df['my_column']` rather than with object notation (`df.my_column`)**. There are many edge cases in which the object notation does not work, along with nuances as to how `pandas` will behave.

In [25]:
diamonds.cut.head()

0      Ideal
1    Premium
2       Good
3    Premium
4       Good
Name: cut, dtype: object

In [ ]:
diamonds['cut'].head()

Remember: This will be a **Series** object, not a **DataFrame**.

<a id='info'></a>

### Examining Your Data With `.info()`

---

When getting acquainted with a new data set, `.info()` should be **the first thing** you examine.

**Types** are very important. They affect the way data will be **represented** in our machine learning models, how data can be joined, whether or not math operators can be applied, and instances in which you can encounter unexpected results.

> _Typical problems that arise when working with new data sets include_:
> - Missing values.
> - Unexpected types (string/object instead of int/float).
> - Dirty data (commas, dollar signs, unexpected characters, etc.).
> - Blank values that are actually "non-null" or single white-space characters.

`.info()` is a function available on every **DataFrame** object. It provides information about:

- The name of the column/variable attribute.
- The type of index (RangeIndex is default).
- The count of non-null values by column/attribute.
- The type of data contained in the column/attribute.
- The unique counts of **dtypes** (`pandas` data types).
- The memory usage of our data set.

#### For example: 

In [26]:
diamonds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   carat    53940 non-null  float64
 1   cut      53940 non-null  object 
 2   color    53940 non-null  object 
 3   clarity  53940 non-null  object 
 4   depth    53940 non-null  float64
 5   table    53940 non-null  float64
 6   price    53940 non-null  int64  
 7   x        53940 non-null  float64
 8   y        53940 non-null  float64
 9   z        53940 non-null  float64
dtypes: float64(6), int64(1), object(3)
memory usage: 4.1+ MB


<a id='describe'></a>

### Summarizing Data with `.describe()`

---

The ```.describe()``` function is useful for taking a quick look at your data. It returns some basic descriptive statistics.

For our example, use the ```.describe()``` function on only the ```carat``` column.

In [27]:
diamonds['carat'].describe()

count    53940.000000
mean         0.797940
std          0.474011
min          0.200000
25%          0.400000
50%          0.700000
75%          1.040000
max          5.010000
Name: carat, dtype: float64

You can also use it on multiple columns, such as ```carat``` and ```price```. They will need to be numeric types. 

In [28]:
diamonds[['carat','price']].describe()

carat         price
count  53940.000000  53940.000000
mean       0.797940   3932.799722
std        0.474011   3989.439738
min        0.200000    326.000000
25%        0.400000    950.000000
50%        0.700000   2401.000000
75%        1.040000   5324.250000
max        5.010000  18823.000000

```.describe()``` gives us the following statistics:

- **Count**, which is equivalent to the number of cells (rows).
- **Mean**, or, the average of the values in the column.
- **Std**, which is the standard deviation.
- **Min**, a.k.a., the minimum value.
- **25%**, or, the 25th percentile of the values.
- **50%**, or, the 50th percentile of the values ( which is the equivalent to the median).
- **75%**, or, the 75th percentile of the values.
- **Max**, which is the maximum value.

<img src="https://snag.gy/AH6E8I.jpg">

#### Summary Functions

There are also built-in math functions that will work on a column of a DataFrame. 

For example, I can use the ```.mean()``` function on the `carat` column to get the mean carat weight for all the diamonds.

In [29]:
diamonds[['carat']].mean()

carat    0.79794
dtype: float64

We can also use them on multiple columns:

In [30]:
diamonds[['carat','price']].max()

carat        5.01
price    18823.00
dtype: float64

if the columns are all numeric you can call them directly on the DataFrame, otherwise you might get a warning:

In [31]:
diamonds.mean()

C:\Users\ng_sm\AppData\Local\Temp/ipykernel_4892/2453091466.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  diamonds.mean()


carat       0.797940
depth      61.749405
table      57.457184
price    3932.799722
x           5.731157
y           5.734526
z           3.538734
dtype: float64

Number of occurrences

Using `value_counts()`, `unique()` and `nunique()`

In [120]:
diamonds.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

In [121]:
diamonds['cut'].value_counts()

Ideal        21551
Premium      13791
Very Good    12082
Good          4906
Fair          1610
Name: cut, dtype: int64

In [122]:
diamonds['cut'].unique()

array(['Ideal', 'Premium', 'Good', 'Very Good', 'Fair'], dtype=object)

In [127]:
diamonds['cut'].nunique()

carat        273
cut            5
color          7
clarity        8
depth        184
table        127
price      11602
x            554
y            552
z            375
dtype: int64

<a id='independent_practice'></a>

### Independent Practice

---

Now that we know a little bit about basic DataFrame use, let's practice on a new data set.

> Pro tip: When your cursor is in a string, you can use the "tab" key to browse file system resources and get a relative reference for the files that can be loaded in Jupyter notebook. Remember, you have to use your arrow keys to navigate the files populated in the UI. 

<img src="https://snag.gy/IlLNm9.jpg">

1. Find and load the `cars` data set into a DataFrame (in the `datasets` directory).
2. Print out the columns.
3. What does the data set look like in terms of dimensions?
4. Check the types of each column.
  a. What is the most common type?
  b. How many entries are there?
  c. How much memory does this data set consume?
5. Examine the summary statistics of the data set.

In [107]:
csv_file = "datasets/cars.csv"
cars = pd.read_csv(csv_file)

In [33]:
# column names
cars.columns

Index(['mpg', 'cyl', 'disp', 'hp', 'drat', 'wt', 'qsec', 'vs', 'am', 'gear',
       'carb'],
      dtype='object')

In [34]:
# shape
cars.shape

(26, 11)

In [35]:
# column data types, number of entries and memory used. seems like alot of info
cars.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26 entries, 0 to 25
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   mpg     26 non-null     float64
 1   cyl     26 non-null     int64  
 2   disp    26 non-null     float64
 3   hp      26 non-null     int64  
 4   drat    26 non-null     float64
 5   wt      26 non-null     float64
 6   qsec    26 non-null     float64
 7   vs      26 non-null     int64  
 8   am      26 non-null     int64  
 9   gear    26 non-null     int64  
 10  carb    26 non-null     int64  
dtypes: float64(5), int64(6)
memory usage: 2.4 KB


In [37]:
# summary stats 
cars.mean()

mpg      19.792308
cyl       6.307692
disp    240.373077
hp      138.730769
drat      3.515385
wt        3.346500
qsec     18.244615
vs        0.461538
am        0.269231
gear      3.423077
carb      2.538462
dtype: float64

<a id='indexing'></a>

### `pandas` Indexing 

---

More often than not, we want to operate on or extract specific portions of our data. When we perform indexing on a DataFrame or Series, we can specify a certain section of the data.

`pandas` has three properties you can use for indexing:

- **`.loc`** indexes with the _labels_ for rows and columns.
- **`.iloc`** indexes with the _integer positions_ for rows and columns.

To help clarify these differences, let's first reset the row labels to letters using the ```.index``` attribute:

In [38]:
new_index_values = ['A','B','C','D','E','F','G','H','I','J','K','L','M',
                    'N','O','P','Q','R','S','T','U','V','W','X','Y','Z']
cars.index=new_index_values


In [39]:
cars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
A  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
B  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
C  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
D  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
E  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

Using the **`.loc`** indexer, we can pull out rows **'B' to 'F'** and the **`hp` and `wt`** columns.

In [51]:
subset = cars.loc[['B','C','D','E','F'], ['mpg','hp','wt']]

In [49]:
subset

mpg   hp     wt
B  21.0  110  2.875
C  22.8   93  2.320
D  21.4  110  3.215
E  18.7  175  3.440
F  18.1  105  3.460

Using a boolean mask to subset the rows

In [50]:
# Find all where the mpg = 21
cars['mpg']==21

A     True
B     True
C    False
D    False
E    False
F    False
G    False
H    False
I    False
J    False
K    False
L    False
M    False
N    False
O    False
P    False
Q    False
R    False
S    False
T    False
U    False
V    False
W    False
X    False
Y    False
Z    False
Name: mpg, dtype: bool

In [52]:
# Using the boolean mask cars['mpg']==21 to choose the rows
cars.loc[cars['mpg']==21,['mpg','hp','wt']]

mpg   hp     wt
A  21.0  110  2.620
B  21.0  110  2.875

We can do the same thing with the **`.iloc`** indexer, but we have to use integers for the location.

In [58]:
subset = cars.iloc[4:10, [3,5]]

In [59]:
subset

hp    wt
E  175  3.44
F  105  3.46
G  245  3.57
H   62  3.19
I   95  3.15
J  123  3.44

If you try to index the rows or columns with integers using **`.loc`**, you will get an error.

Note that you can automatically reorder the data just by reordering the indices you enter when you perform the indexing operation!

While we created an index earlier, we can also use a column to set an index.

In [62]:
cars.index = cars['mpg']

cars.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
mpg                                                                
21.0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
21.0  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
22.8  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
21.4  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
18.7  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

Is mpg the best feature to use as an index?  

If it isn't we can use the `df.reset_index()` to reset our index.

In [69]:
cars.reset_index(inplace=True)
cars.head()

level_0  index   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0        0      0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1        1      1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2        2      2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3        3      3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4        4      4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

<a id='creating_dataframes'></a>

### Creating DataFrames

---

The simplest way to create your own DataFrame without importing data from a file is to give the ```pd.DataFrame()``` instantiator a dictionary.

In [70]:
mydata = pd.DataFrame({'Letters':['A','B','C'], 'Integers':[1,2,3], 'Floats':[2.2, 3.3, 4.4]})

In [71]:
mydata

Letters  Integers  Floats
0       A         1     2.2
1       B         2     3.3
2       C         3     4.4

As you might expect, the dictionary needs to have lists of values that are all the same length. The keys correspond to the names of the columns, and the values correspond to the data in the columns.

<a id='dtypes'></a>

### Examining Data Types

---

`pandas` comes with a useful property for looking solely at the data types of your DataFrame columns. Use ```.dtypes``` on your DataFrame:

In [72]:
mydata.dtypes

Letters      object
Integers      int64
Floats      float64
dtype: object

This will show you the data type of each column. Strings are stored as a type called "object," as they are not guaranteed to take up a set amount of space (strings can be any length).

<a id='renaming_assignment'></a>

### Renaming and Assignment

---

`pandas` makes it easy to change column names and assign values to your DataFrame.

Say, for example, we want to change the column name `Integers` to `int`:

In [73]:
mydata.rename(columns={mydata.columns[1]:'int'}, inplace=True) # inplace = True updates mydata
print(mydata.columns)

Index(['Letters', 'int', 'Floats'], dtype='object')


In [74]:
mydata

Letters  int  Floats
0       A    1     2.2
1       B    2     3.3
2       C    3     4.4

If you want to change every column name, you can just assign a new list to the ```.columns``` property.

In [75]:
mydata.columns = ['A','B','C']
print(mydata.head())

   A  B    C
0  A  1  2.2
1  B  2  3.3
2  C  3  4.4


<a id='filtering'></a>

### Filtering Logic

---

One of the most powerful features of DataFrames is the ability to use logical commands to filter data.

Subset the ```diamonds``` data for only the rows in which `price` is greater than 7000.

In [81]:
diamonds[diamonds['price'] > 7000]

carat        cut color clarity  depth  table  price     x     y     z
17457   1.01  Very Good     D     VS2   60.6   56.0   7001  6.45  6.51  3.93
17458   1.01  Very Good     F     VS1   63.7   56.0   7001  6.27  6.33  4.01
17459   1.01       Good     F    VVS2   65.9   54.0   7001  6.18  6.27  4.10
17460   1.30    Premium     D     SI2   60.2   59.0   7002  7.00  7.05  4.23
17461   1.00  Very Good     E     VS1   61.2   57.0   7002  6.43  6.47  3.95
...      ...        ...   ...     ...    ...    ...    ...   ...   ...   ...
27745   2.00  Very Good     H     SI1   62.8   57.0  18803  7.95  8.00  5.01
27746   2.07      Ideal     G     SI2   62.5   55.0  18804  8.20  8.13  5.11
27747   1.51      Ideal     G      IF   61.7   55.0  18806  7.37  7.41  4.56
27748   2.00  Very Good     G     SI1   63.5   56.0  18818  7.90  7.97  5.04
27749   2.29    Premium     I     VS2   60.8   60.0  18823  8.50  8.47  5.16

[9273 rows x 10 columns]

In [76]:
# boolean mask that was used
diamonds['price'] > 7000

0        False
1        False
2        False
3        False
4        False
         ...  
53935    False
53936    False
53937    False
53938    False
53939    False
Name: price, Length: 53940, dtype: bool

#### Filtering on Multiple Conditions

We can also filter on _multiple conditions_. 
The format for multiple conditions is:

`df[ (df['col1'] == value1) & (df['col2'] == value2) ]`

`df[ (df['col1'] == value1) | (df['col2'] == value2) ]`
Or, more simply:

`df[ (CONDITION 1) & (CONDITION 2) ]`

`df[ (CONDITION 1) | (CONDITION 2) ]`

Which eventually may evaluate to something like:

`df[ True & False ]`

...on a row-by-row basis. If the end result is `False`, the row is omitted.

_Don't forget parentheses in your conditions!_ This is a common mistake.

#### Example 
Subset the data for `price` greater than 7000 like before, but now, also include where the `cut` is 'Ideal'.

In [82]:
diamonds[(diamonds['price'] > 7000) & (diamonds['cut']== 'Ideal')]

carat    cut color clarity  depth  table  price     x     y     z
17462   1.21  Ideal     I    VVS2   62.7   56.0   7002  6.80  6.83  4.27
17471   1.01  Ideal     E     VS2   61.9   57.0   7013  6.43  6.37  3.96
17472   1.01  Ideal     F     VS2   61.9   54.0   7014  6.45  6.49  4.00
17473   1.04  Ideal     F     VS1   61.4   57.0   7015  6.51  6.56  4.01
17474   1.34  Ideal     I     VS1   62.2   57.0   7016  7.02  7.06  4.38
...      ...    ...   ...     ...    ...    ...    ...   ...   ...   ...
27735   1.60  Ideal     F     VS1   62.0   56.0  18780  7.47  7.52  4.65
27738   2.05  Ideal     G     SI1   61.9   57.0  18787  8.10  8.16  5.03
27741   2.15  Ideal     G     SI2   62.6   54.0  18791  8.29  8.35  5.21
27746   2.07  Ideal     G     SI2   62.5   55.0  18804  8.20  8.13  5.11
27747   1.51  Ideal     G      IF   61.7   55.0  18806  7.37  7.41  4.56

[3266 rows x 10 columns]

In [ ]:
# What about diamonds where the cut is Premium OR the carat weight is greater than 0.50?
# "Or" logic - use pipe (|)

In [83]:
diamonds[(diamonds['cut']=='Premium')|(diamonds['carat']>0.5)]

carat        cut color clarity  depth  table  price     x     y     z
1       0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
3       0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
12      0.22    Premium     F     SI1   60.4   61.0    342  3.88  3.84  2.33
14      0.20    Premium     E     SI2   60.2   62.0    345  3.79  3.75  2.27
15      0.32    Premium     E      I1   60.9   58.0    345  4.38  4.42  2.68
...      ...        ...   ...     ...    ...    ...    ...   ...   ...   ...
53935   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75  5.76  3.50
53936   0.72       Good     D     SI1   63.1   55.0   2757  5.69  5.75  3.61
53937   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66  5.68  3.56
53938   0.86    Premium     H     SI2   61.0   58.0   2757  6.15  6.12  3.74
53939   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83  5.87  3.64

[39323 rows x 10 columns]

#### Calculations on filtered data

Let's calculate the **mean carat weight** for diamonds with `cut` **Premium**
> Think: What are the component parts of this problem?

In [94]:
# First find the diamonds where the cut == 'Premium'
premium_diamonds = diamonds[diamonds['cut']=='Premium']

# Then select the 'carat' column as a series
carat_series = premium_diamonds['carat']

# Finally, find the mean

carat_series.mean()

0.8919548981219524

In [95]:
diamonds[diamonds['cut']=='Premium']['carat'].mean()

0.8919548981219524

In [96]:
# answer in one line
mean_carat = diamonds[diamonds['cut']=='Premium']['carat'].mean()
print(f'{mean_carat:.2f}')

0.89


<a id='sorting'></a>

### Sorting

We can sort the DataFrame by Series, or by the entire DataFrame by specifying which columnd to sort. 

In [97]:
# We can sort individual Series...
cars['mpg'].sort_values().head()

15    10.4
14    10.4
23    13.3
6     14.3
16    14.7
Name: mpg, dtype: float64

In [98]:
# sort in descending order
cars['mpg'].sort_values(ascending=False).head()

19    33.9
17    32.4
18    30.4
25    27.3
7     24.4
Name: mpg, dtype: float64

In [99]:
# Sort the entire DataFrame by the specific column
cars.sort_values('mpg').head()

level_0  index   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
15       15     15  10.4    8  460.0  215  3.00  5.424  17.82   0   0     3   
14       14     14  10.4    8  472.0  205  2.93  5.250  17.98   0   0     3   
23       23     23  13.3    8  350.0  245  3.73  3.840  15.41   0   0     3   
6         6      6  14.3    8  360.0  245  3.21  3.570  15.84   0   0     3   
16       16     16  14.7    8  440.0  230  3.23  5.345  17.42   0   0     3   

    carb  
15     4  
14     4  
23     4  
6      4  
16     4

In [102]:
# Or the entire DataFrame by more than one column using lists
# cars.sort_values(by=??? , ascending=??? ).head()
cars.sort_values(by=['cyl','carb','gear'], ascending=[False,True,True])

level_0  index   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
4         4      4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   
21       21     21  15.5    8  318.0  150  2.76  3.520  16.87   0   0     3   
22       22     22  15.2    8  304.0  150  3.15  3.435  17.30   0   0     3   
24       24     24  19.2    8  400.0  175  3.08  3.845  17.05   0   0     3   
11       11     11  16.4    8  275.8  180  3.07  4.070  17.40   0   0     3   
12       12     12  17.3    8  275.8  180  3.07  3.730  17.60   0   0     3   
13       13     13  15.2    8  275.8  180  3.07  3.780  18.00   0   0     3   
6         6      6  14.3    8  360.0  245  3.21  3.570  15.84   0   0     3   
14       14     14  10.4    8  472.0  205  2.93  5.250  17.98   0   0     3   
15       15     15  10.4    8  460.0  215  3.00  5.424  17.82   0   0     3   
16       16     16  14.7    8  440.0  230  3.23  5.345  17.42   0   0     3   
23       23     23  13.3    8  350.0  245  3.73  3.840  15.41   0   0     3   
3         3      3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
5         5      5  18.1    6  225.0  105  2.76  3.460  20.22   1   0     3   
0         0      0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1         1      1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
9         9      9  19.2    6  167.6  123  3.92  3.440  18.30   1   0     4   
10       10     10  17.8    6  167.6  123  3.92  3.440  18.90   1   0     4   
20       20     20  21.5    4  120.1   97  3.70  2.465  20.01   1   0     3   
2         2      2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
17       17     17  32.4    4   78.7   66  4.08  2.200  19.47   1   1     4   
19       19     19  33.9    4   71.1   65  4.22  1.835  19.90   1   1     4   
25       25     25  27.3    4   79.0   66  4.08  1.935  18.90   1   1     4   
7         7      7  24.4    4  146.7   62  3.69  3.190  20.00   1   0     4   
8         8      8  22.8    4  140.8   95  3.92  3.150  22.90   1   0     4   
18       18     18  30.4    4   75.7   52  4.93  1.615  18.52   1   1     4   

    carb  
4      2  
21     2  
22     2  
24     2  
11     3  
12     3  
13     3  
6      4  
14     4  
15     4  
16     4  
23     4  
3      1  
5      1  
0      4  
1      4  
9      4  
10     4  
20     1  
2      1  
17     1  
19     1  
25     1  
7      2  
8      2  
18     2

## Independent Practice

With our cars dataset already loaded, let's explore our dataset a bit more thoroughly to gain some familiarity with beginning exploratory analysis.

### 1. Select only data for "cars" when "hp" is "150-200".

In [108]:
cars[(cars['hp']>=150) &(cars['hp']<=200) ]

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
4   18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2
11  16.4    8  275.8  180  3.07  4.070  17.40   0   0     3     3
12  17.3    8  275.8  180  3.07  3.730  17.60   0   0     3     3
13  15.2    8  275.8  180  3.07  3.780  18.00   0   0     3     3
21  15.5    8  318.0  150  2.76  3.520  16.87   0   0     3     2
22  15.2    8  304.0  150  3.15  3.435  17.30   0   0     3     2
24  19.2    8  400.0  175  3.08  3.845  17.05   0   0     3     2

### 2. Select only rows with index 5-10, for variables / columns "disp" and "hp"

In [113]:
# which function shall we use
# loc or iloc
cars.loc[5:10,['disp','hp']]

disp   hp
5   225.0  105
6   360.0  245
7   146.7   62
8   140.8   95
9   167.6  123
10  167.6  123

### 3. Select the columns by numeric offset 2-5, rows with numeric index 3-7

In [116]:
cars.iloc[3:8,2:6]

disp   hp  drat     wt
3  258.0  110  3.08  3.215
4  360.0  175  3.15  3.440
5  225.0  105  2.76  3.460
6  360.0  245  3.21  3.570
7  146.7   62  3.69  3.190

### 4. Find the mean of `hp`


In [117]:
cars['hp'].mean()

138.73076923076923

### 4. Find the mean of `hp` for cars with more then 4 cylinders

In [118]:
cars[cars['cyl']>4]['hp'].mean()

167.27777777777777

### 5. Sort the `cars` DataFrame by `cyl` in ascending order, then `mpg` in descending order

In [119]:
cars.sort_values(['cyl','mpg'],ascending=[True, False])

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
19  33.9    4   71.1   65  4.22  1.835  19.90   1   1     4     1
17  32.4    4   78.7   66  4.08  2.200  19.47   1   1     4     1
18  30.4    4   75.7   52  4.93  1.615  18.52   1   1     4     2
25  27.3    4   79.0   66  4.08  1.935  18.90   1   1     4     1
7   24.4    4  146.7   62  3.69  3.190  20.00   1   0     4     2
2   22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
8   22.8    4  140.8   95  3.92  3.150  22.90   1   0     4     2
20  21.5    4  120.1   97  3.70  2.465  20.01   1   0     3     1
3   21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
0   21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1   21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
9   19.2    6  167.6  123  3.92  3.440  18.30   1   0     4     4
5   18.1    6  225.0  105  2.76  3.460  20.22   1   0     3     1
10  17.8    6  167.6  123  3.92  3.440  18.90   1   0     4     4
24  19.2    8  400.0  175  3.08  3.845  17.05   0   0     3     2
4   18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2
12  17.3    8  275.8  180  3.07  3.730  17.60   0   0     3     3
11  16.4    8  275.8  180  3.07  4.070  17.40   0   0     3     3
21  15.5    8  318.0  150  2.76  3.520  16.87   0   0     3     2
13  15.2    8  275.8  180  3.07  3.780  18.00   0   0     3     3
22  15.2    8  304.0  150  3.15  3.435  17.30   0   0     3     2
16  14.7    8  440.0  230  3.23  5.345  17.42   0   0     3     4
6   14.3    8  360.0  245  3.21  3.570  15.84   0   0     3     4
23  13.3    8  350.0  245  3.73  3.840  15.41   0   0     3     4
14  10.4    8  472.0  205  2.93  5.250  17.98   0   0     3     4
15  10.4    8  460.0  215  3.00  5.424  17.82   0   0     3     4

<a id='review'></a>

### Review 

We covered a lot of ground! It's ok if this takes a while to gel.

```python

# basic DataFrame operations
df.head()
df.tail()
df.shape
df.columns
df.Index
df.info()

# selecting columns
df.column_name
df['column_name']
df[['column_name1','column_name2','column_name3']]

# notable columns operations
df.describe() # five number summary
df['col1'].nunique() # number of unique values
df['col1'].value_counts() # number of occurrences of each value in column

# filtering
df[ df['col1'] < 50 ] # filter column to be less than 50
df[ (df['col1'] == value1) & (df['col2'] > value2) ] # filter column where col1 is equal to value1 AND col2 is greater to value 2

# sorting
df.sort_values(by='column_name', ascending = False) # sort biggest to smallest

```


It's common to refer back to your own code *all the time.* Don't hesitate to reference this guide! 🐼
